<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [131]:
import pandas as pd
import psycopg2

In [132]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [133]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [182]:
query3_1 = f'''select count(*)
            from vacancies
        '''
df = pd.read_sql_query(query3_1, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/1707171897.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query3_1, connection)


count
0  49197

В базе данных 49197 вакансии

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [183]:
query3_2 = f'''select count(*)
            from employers
        '''
df = pd.read_sql_query(query3_2, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/3410260031.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query3_2, connection)


count
0  23501

В базе данных 23501 работодателей, что в более чем в 2 раза меньше, чем количество вакансий. Что значит, что в большинстве случае работодатель имеет по несколько вакансий.

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [184]:
query3_3 = f'''select count(*)
            from areas
        '''
df = pd.read_sql_query(query3_3, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/161713720.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query3_3, connection)


count
0   1362

В базе данных 1362 региона, некторорые работодатели представлены в нескольких регионах.

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [177]:
query3_4 = f'''select count(*)
            from industries
        '''
df = pd.read_sql_query(query3_4, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/743582147.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    294

В базе данных 294 индустрии

В представленной базе данных содержится 49197 вакансий от 23501 работадателей из 1362 регионов, которые представляют 294 индустрии.

***

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [185]:
query4_1 = f'''select a.name, count(v.area_id) as cnt
            from vacancies v
            join areas a on v.area_id = a.id
            group by v.area_id, a.name
            order by 2 DESC
            limit 5
        '''
df = pd.read_sql_query(query4_1, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/2004434097.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query4_1, connection)


name   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

Лидеры по количеству вакансий - 5 городов: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. Основное количество вакансий представлены в столицах стран или больших городах.

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [186]:
query4_2 = f'''select count(*)
            from vacancies v
            where not salary_from is null
            or not salary_to is null
        '''
df = pd.read_sql_query(query4_2, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/2901952598.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query4_2, connection)


count
0  24073

В 24073 вакансиях указано значение заработной платы, что составляет менее чем 50% от всех вакансий. Можно сделать вывод, что большинство компаний предпочитает обсуждать заработную плату с кандидатом, не выставляя конкретной цифры в описании вакансии.

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [187]:
query4_3 = f'''select ROUND(AVG(salary_from), 0) as mean_salary_from, 
            ROUND(AVG(salary_to), 0) as mean_salary_to
            from vacancies v
        '''
df = pd.read_sql_query(query4_3, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/2889325311.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query4_3, connection)


mean_salary_from  mean_salary_to
0           71065.0        110537.0

Cредние значения для нижней границы зарплатной вилки равно 71065, верхней границы зарплатной вилки равно 110537.

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [188]:
query4_4 = f'''select schedule, count(schedule), employment
            from vacancies v
            group by schedule, employment
            order by 2 DESC
            offset 1 limit 1
        '''
df = pd.read_sql_query(query4_4, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/3049570585.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query4_4, connection)


schedule  count        employment
0  Удаленная работа   7802  Полная занятость

Удалённая работа — Полная занятость занимает 2 место, что означает большую готовность компаний предоставлять вакансии с удаленной работой.

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [189]:
query4_5 = f'''select experience, count(*)
            from vacancies
            group by experience
            order by count(*)
        '''
df = pd.read_sql_query(query4_5, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/2700463298.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

В основной доле вакансий, примерно в 53%, требуется небольшой опыт от 1 до 3 лет, на втором месте по количеству, примерно 29%, опыт от 3 до 6 лет. 15% вакансий не требуют опыта от кандидата, что является неплохим процентом. Всего 3% работодателей ищут кандидатов с опытом более 6 лет, что говорит о том, что предпочтечение работодателей отдается молодым специалистам.

Вывод: основное количество вакансий распределено в крупных городах, в средняя заработная плата, указанная в представленных вакансиях от 71065 руб. до 110537 руб, с учетом того что заработная плата указана в менее чем 50% случаях. Также основная доля вакансий требует опыт работы от 1 до 3 лет.

***

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [190]:
query5_1 = f'''(select e.name, count(*)
            from vacancies v
            join employers e on v.employer_id = e.id
            group by e.name
            order by 2 DESC
            limit 1)
            union all
            (select e.name, count(*)
            from vacancies v
            join employers e on v.employer_id = e.id
            group by e.name
            order by 2 DESC
            offset 4 
            limit 1)
        '''
df = pd.read_sql_query(query5_1, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/1179034331.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_1, connection)


name  count
0         Яндекс   1933
1  Газпром нефть    331

4% из всех представленных вакансий трубуются в компанию Яндекс, данная компания занимает 1 место по количеству вакансий. 5 место по количеству вакансий занимает компания Газпром нефть. Количество вакансий зависит от размера компании.

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [191]:
query5_2 = f'''(select a.name, count(v.id), count(e.id)
            from areas a
            left join vacancies v on v.area_id = a.id
            left join employers e on e.area = a.id
            where v.id is null
            and not e.id is null
            group by a.id
            order by 3 DESC
            limit 1)
        '''
df = pd.read_sql_query(query5_2, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/859544887.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_2, connection)


name  count  count
0  Россия      0    410

Работодатели, указывающие регион Россия, не имеют вакансий. Это показывает, что записи с регионом "страна", некорректы для дальнейшего анализа.

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [192]:
query5_3 = f'''(select e.name, count(distinct a.id)
            from vacancies v
            left join areas a on v.area_id = a.id
            left join employers e on v.employer_id = e.id
            group by e.name
            order by 2 DESC
            limit 1
            )
        '''
df = pd.read_sql_query(query5_3, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/209878418.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_3, connection)


name  count
0  Яндекс    181

Максимальное количество регионов, представленных вакансий имеет компания Яндекс - 181 регион.

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [193]:
query5_4 = f'''(select count(*)
            from employers e
            left join employers_industries ei on ei.employer_id = e.id
            where ei.industry_id is null
            )
        '''
df = pd.read_sql_query(query5_4, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/17442584.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_4, connection)


count
0   8419

36% работодателей не указывют свою сферу деятельности при оформление вакансий.

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [194]:
query5_5 = f'''(select e.name
            from employers e
            left join employers_industries ei on ei.employer_id = e.id
            group by e.name
            having count(distinct ei.industry_id) = 4
            order by 1
            offset 2 limit 1
            )
        '''
df = pd.read_sql_query(query5_5, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/2885300682.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_5, connection)


name
0  2ГИС

Третье место в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности занимает компания 2ГИС.

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [195]:
query5_6 = f'''(select count(*)
            from employers e
            left join employers_industries ei on ei.employer_id = e.id
            left join industries i on ei.industry_id = i.id
            where i.name = 'Разработка программного обеспечения'
            )
        '''
df = pd.read_sql_query(query5_6, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/3538068853.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_6, connection)


count
0   3553

15% работодателей - 3553, указывают в сфере деятельности "Разработка ПО"

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [196]:
query5_7 = f'''(select a.name, count(v.id)
            from vacancies v
            left join employers e on v.employer_id = e.id
            left join areas a on v.area_id = a.id
            where e.name = 'Яндекс' 
            and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
            group by a.name
            union all
            select 'Total', count(v.id)
            from vacancies v
            left join employers e on v.employer_id = e.id
            left join areas a on v.area_id = a.id
            where e.name = 'Яндекс' 
            and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
            order by 2 DESC
            )
        '''
df = pd.read_sql_query(query5_7, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/3603638735.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_7, connection)


name  count
0             Total    485
1            Москва     54
2   Санкт-Петербург     42
3      Екатеринбург     39
4   Нижний Новгород     36
5       Новосибирск     35
6           Воронеж     32
7         Краснодар     30
8            Самара     26
9               Уфа     26
10            Пермь     25
11           Казань     25
12   Ростов-на-Дону     25
13        Волгоград     24
14       Красноярск     23
15        Челябинск     22
16             Омск     21

25% вакансии от всех вакансий представленных в компании Яндекс, находятся в городах миллионниках.

***

Выводы: большинство вакансий представленно в крупных компаниях, которые имеют преставительства в нескольких регионах, в основнов в городах-миллионниках. Крупные компании имеют несколько сфер деятельности, что удобно для соискателей вакансий. Однако 38% компаний не заполняют сферу деятельности. 15% компаний занимаются сферой "Разработка ПО", что актуально для IT специалистов.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [197]:
query6_1= f'''(select count(name)
            from vacancies
            where name ilike('%data%')
            or name ilike('%данн%')
            )
        '''
df = pd.read_sql_query(query6_1, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/4229670932.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query6_1, connection)


count
0   1771

1771 вакансий, примерно 4%, имеют отношение к данным.

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [198]:
query6_2 = f'''(select count(*)
            from vacancies
            where (name ilike ('%data scientist%')
            or name ilike ('%data science%')
            or name ilike ('%исследователь данных%')
            or (name like ('%ML%') and not name ilike('%HTML%'))
            or name ilike('%machine learning%')
            or name ilike('%машинн%обучен%'))
            and (name ilike('%junior%')
            or experience = 'Нет опыта'
            or employment = 'Стажировка')
            )
        '''
df = pd.read_sql_query(query6_2, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/2024065959.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query6_2, connection)


count
0     51

Достаточно небольшое количество вакансий для начинающих дата-сайентистов представлено в бд - 51 вакансия.

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [205]:
query6_3 = f'''(select count(*)
            from vacancies
            where (key_skills ilike ('%sql%') or key_skills ilike ('%postgres%'))
            and (name ilike ('%data scientist%')
            or name ilike ('%data science%')
            or name ilike ('%исследователь данных%')
            or (name like ('%ML%') and not name ilike('%HTML%'))
            or name ilike('%machine learning%')
            or name ilike('%машинн%обучен%'))
            )
        '''
df = pd.read_sql_query(query6_3, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/3796087825.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query6_3, connection)


count
0    201

В 201 вакансиях для дата-сайентистов в целом, требуется навык работы с SQL.

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [200]:
query6_4 = f'''(select count(*)
            from vacancies
            where (key_skills ilike ('%Python%'))
            and (name ilike ('%data scientist%')
            or name ilike ('%data science%')
            or name ilike ('%исследователь данных%')
            or (name like ('%ML%') and not name ilike('%HTML%'))
            or name ilike('%machine learning%')
            or name ilike('%машинн%обучен%'))
            )
        '''
df = pd.read_sql_query(query6_4, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/4090375082.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query6_4, connection)


count
0    351

В 351 вакансии для дата-сайентистов в целом, требуется навык работы с Python.

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [210]:
query6_5 = f'''(select ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1), 0)
            from vacancies
            where (not key_skills is null)
            and (name ilike ('%data scientist%')
            or name ilike ('%data science%')
            or name ilike ('%исследователь данных%')
            or (name like ('%ML%') and not name ilike('%HTML%'))
            or name ilike('%machine learning%')
            or name ilike('%машинн%обучен%'))
            )
        '''
df = pd.read_sql_query(query6_5, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/1499980039.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query6_5, connection)


round
0    6.0

В среднем для дата-сайентистов трубется 6 ключевых навыков

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [170]:
query6_6 = f'''( select ROUND(AVG(salary), 0) 
                from(
                ((select (salary_from + salary_to) / 2 as salary
                    from vacancies
                    where not salary_from is null
                    and not salary_to is null
                    and (name ilike ('%data scientist%')
                    or name ilike ('%data science%')
                    or name ilike ('%исследователь данных%')
                    or (name like ('%ML%') and not name ilike('%HTML%'))
                    or name ilike('%machine learning%')
                    or name ilike('%машинн%обучен%'))
                    and experience = 'От 3 до 6 лет')
                    Union all
                    (select salary_from
                        from vacancies
                        where not salary_from is null
                        and salary_to is null
                        and (name ilike ('%data scientist%')
                        or name ilike ('%data science%')
                        or name ilike ('%исследователь данных%')
                        or (name like ('%ML%') and not name ilike('%HTML%'))
                        or name ilike('%machine learning%')
                        or name ilike('%машинн%обучен%'))
                        and experience = 'От 3 до 6 лет')
                        Union all
                        (select salary_to
                        from vacancies
                        where  salary_from is null
                        and not salary_to is null
                        and (name ilike ('%data scientist%')
                        or name ilike ('%data science%')
                        or name ilike ('%исследователь данных%')
                        or (name like ('%ML%') and not name ilike('%HTML%'))
                        or name ilike('%machine learning%')
                        or name ilike('%машинн%обучен%'))
                        and experience = 'От 3 до 6 лет'))) as list_salary
            )
        '''
df = pd.read_sql_query(query6_6, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/3964033354.py:40: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


round
0  243115.0

Средняя заработная плата для дата-сайентистов с опытом работы 3-6 лет, составляет 243115 руб.

Выводы: примерно 4% из представленных вакансий имею отношение к данным, из них всего 51 вакансия предназначены для начинающего дата-сайентиста без опыта работы. Из всех вакансий для дата-сайнтистов, не учитывая опыт работы - 480 вакансий, в примерно 50% случаях требуется знание SQL и в более чем 50% случаях требуются навыки работы с Python. В среднем, чтобы работать дата-сайнтистом необходимо 6 ключевых навыков. А средняя заработная плата опытного дата-сайентиста, с опытом 3-6 лет составляет 243115 руб, что больше средней заработной платы по всем вакансиям в примерно 2 раза, что доказывает востребовательность и высокоплачиваемость специалистов data-science.

***

# Общий вывод по проекту

С помощью данного исследования мы узнали:
*  Основаная доля вакансий сосредоточена в больших компаниях, расположенных в крупных городах или городах- миллионниках.
* Крупные компании имеют несколько сфер деятельности, что позволяет сотрудникам развить более гибкие навыки и получить опыт из разных сфер.
* Примерно 4% от всех вакансий относятся данным, что является не маленьким показателем, учитывая разносторонний рынок вакансий.
* Вакансия data-scientist является достаточно востребованной и высокоплачиваемой.

 Для дополнительного анализа, можем детально рассмотреть топ 5 индустрий и сделать прогноз по актуальности профессий в ближайшее время.

In [213]:
query_add = f'''(select i.name, count(*)
            from employers e
            left join employers_industries ei on ei.employer_id = e.id
            left join industries i on ei.industry_id = i.id
            where not i.name is null
            group by i.name
            order by 2 DESC
            limit 5
            )
        '''
df = pd.read_sql_query(query_add, connection)
df

/var/folders/g1/qmxhdvlx3kx97plxw0bx1vnw0000gn/T/ipykernel_20508/2807233672.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_add, connection)


name  count
0                Разработка программного обеспечения   3553
1  Системная интеграция,  автоматизации технологи...   2993
2  Интернет-компания (поисковики, платежные систе...   1675
3  Маркетинговые, рекламные, BTL, дизайнерские, E...    798
4                              Консалтинговые услуги    662

Вывод: исходя из полученных данных видно, что первые 3 места занимают индустрии, которые можно отнести к IT, также преобладают индустрии маркетинг и консалтинг.
Можно предположить, в связи с достаточной потребностью рынка в вакансиях IT, маркетинг и консалтинг, эти сферы будут популярны в ближайшем будущем. 
